pkg/provisioner/ironic/ironic.go

```diff
@@ -1295,3 +1347,67 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+// Retrieve configuration in form of Ironic's clean steps
+func (p *ironicProvisioner) buildProvisionCleanSteps(logicalDisks []nodes.LogicalDisk) (cleanSteps []nodes.CleanStep) {
+	// Buil clean steps for RAID configuration
+	raidCleanSteps := p.buildRAIDCleanSteps(logicalDisks)
+	cleanSteps = append(cleanSteps, raidCleanSteps...)
+	return
+}
+
+// A private method to build the clean steps for RAID configuration from BaremetalHost spec
+func (p *ironicProvisioner) buildRAIDCleanSteps(logicalDisks []nodes.LogicalDisk) (cleanSteps []nodes.CleanStep) {
+	// ‘create_configuration’ doesn’t remove existing disks. It is recommended
+	// to add ‘delete_configuration’ before ‘create_configuration’ to make sure
+	// that only the desired logical disks exist in the system after manual cleaning.
+	raidCleanSteps := []nodes.CleanStep{{Interface: "raid", Step: "delete_configuration"}}
+	if len(logicalDisks) > 0 {
+		raidCleanSteps = append(
+			raidCleanSteps,
+			nodes.CleanStep{Interface: "raid", Step: "create_configuration"},
+		)
+	}
+	return
+}
+
+// A private method to build RAID disks
+func (p *ironicProvisioner) buildTargetRAIDCfg(
+	raidConfig *metal3v1alpha1.RAIDConfig) (logicalDisks []nodes.LogicalDisk){
+	isRootVol := raidConfig.RootVolume != nil
+	if isRootVol {
+		rootVol := raidConfig.RootVolume
+		rootDisk := nodes.LogicalDisk{
+			RAIDLevel:             rootVol.Level,
+			VolumeName:            rootVol.Name,
+			IsRootVolume:          &isRootVol,
+		}
+		if rootVol.SizeGb != 0 {
+			rootDisk.SizeGB = &(rootVol.SizeGb)
+		}
+		if rootVol.DiskType != nil {
+			rootDisk.DiskType = *rootVol.DiskType
+		}
+		if rootVol.NumberOfPhysicalDisks != nil {
+			rootDisk.NumberOfPhysicalDisks = *rootVol.NumberOfPhysicalDisks
+		}
+		logicalDisks = append(logicalDisks, rootDisk)
+	}
+	for _, volume := range raidConfig.Volumes {
+		disk := nodes.LogicalDisk{
+			RAIDLevel:             volume.Level,
+			VolumeName:            volume.Name,
+		}
+		if volume.SizeGb != 0 {
+			disk.SizeGB = &(volume.SizeGb)
+		}
+		if volume.DiskType != nil {
+			disk.DiskType = *volume.DiskType
+		}
+		if volume.NumberOfPhysicalDisks != nil {
+			disk.NumberOfPhysicalDisks = *volume.NumberOfPhysicalDisks
+		}
+		logicalDisks = append(logicalDisks, disk)
```

Comment:
Should we be checking that the volume name is not the same as the root volume that we already added?